src/Shell/WikidataLanguageNamesShell.php

```diff
@@ -0,0 +1,209 @@
+ *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+ *  GNU Affero General Public License for more details.
+ *
+ *  You should have received a copy of the GNU Affero General Public License
+ *  along with this program.  If not, see <http://www.gnu.org/licenses/>.
+ */
+namespace App\Shell;
+
+use App\Lib\LanguagesLib;
+use Cake\Console\Shell;
+use Cake\Core\Configure;
+
+class WikidataLanguageNamesShell extends Shell {
+
+    private $po_files = array();
+    private $tatoeba_languages;
+    private $english_translations;
+
+    private function list_english_translations() {
+        $english_translations = array();
+        foreach ($this->get_localized_translations('en') as $iso_code => $lang_in_english) {
+            $tatoeba_name = $this->tatoeba_languages[$iso_code];
+            if ($tatoeba_name != $lang_in_english) {
+                printf("ISO code \"%s\" is called \"%s\" by Tatoeba, but \"%s\" on Wikidata. Skipping translation.\n",
+                    $iso_code,
+                    $tatoeba_name,
+                    $lang_in_english);
+                continue;
+            }
+            $english_translations[$iso_code] = $lang_in_english;
+        }
+
+        $this->english_translations = $english_translations;
+    }
+
+    private function strip_extension($filename) {
+        return substr($filename, 0, strrpos($filename, '.'));
+    }
+
+    private function add_po_file_or_dir($path) {
+        if (is_file($path)) {
+            $this->add_po_file($path);
+        } elseif(is_dir($path)) {
+            $this->add_po_files_from_dir($path);
+        } else {
+            echo "Can't open '$path'.\n";
+            $this->die_usage();
+        }
+    }
+
+    private function add_po_file($filename) {
+        if (pathinfo($filename, PATHINFO_BASENAME) == 'languages.po') {
+            $dir = pathinfo($filename, PATHINFO_DIRNAME);
+            $code = pathinfo($dir, PATHINFO_BASENAME);
+            $this->po_files[$code] = $filename;
+        }
+    }
+
+    private function add_po_files_from_dir($dir) {
+        $files = scandir($dir);
+        foreach($files as $filename) {
+            if($filename[0] == '.') continue; // Ignore if special or hidden.
+            $this->add_po_file_or_dir($dir.'/'.$filename);
+        }
+    }
+
+    private function get_tatoeba_languages() {
+        Configure::write('Config.language', 'eng');
+        $this->tatoeba_languages = LanguagesLib::languagesInTatoeba();
+    }
+
+    private function get_localized_translations($locale_id) {
+        $query = "
+            SELECT DISTINCT ?iso_code ?name
+            WHERE {
+              ?language wdt:P220 ?iso_code;
+                        rdfs:label ?name;
+              FILTER(lang(?name) = '$locale_id')
+              VALUES ?iso_code {";
+        foreach ($this->tatoeba_languages as $iso_code => $lang_in_english) {
+            $query .= "'$iso_code'";
+        }
```

Comment:
Just suggesting:
```php
implode(array_map(function($l) {return "'$l'";}, array_keys($this->tatoeba_languages)));
```

Reply:
Thanks. I'm just learning php as I go along, so I don't know all those handy functions. On the other hand, the verbosity of creating an anonymous function doesn't really make this read nicely. I think I'll keep the imperative version.

Reply:
As you wish. :+1: